### ISSUE TITLE
Support Kafka Connect permissions at connector level

### ISSUE THREAD

**Issue by Dugong42** [kafbat/kafka-ui#614]
### Issue submitter TODO list

- [X] I've searched for an already existing issues [here](https://github.com/kafbat/kafka-ui/issues)
- [X] I'm running a supported version of the application which is listed [here](https://github.com/kafbat/kafka-ui/blob/main/.github/SECURITY.md) and the feature is not present there

### Is your proposal related to a problem?

Current Kafbat UI permissions related to Kafka Connect only allow a granularity at the level of the Connect cluster. This means any permission covers all connectors in that cluster with no exception.

### Describe the feature you're interested in

I would like to be able to set permissions targeting connectors based on their name for example, like it is done with topics and consumers.

### Version you're running

1.0.0

**Comment by Dugong42** [comment]
The documentation indicates that `selectors` may be implemented in the future, with a proposed config spec.

```yaml
        - resource: connect
          value: "local"
          actions: [ view, edit, create ]
        # connectors selector not implemented yet, use connects
        #      selector:
        #        connector:
        #          name: ".*"
        #          class: 'io.kafbat.connectorName'
```

It tries to implement generic selectors but I don't see how that would work in the general case. Perhaps the `name` of the selector should actually be a generic `value` instead. Also the class may be different in frontend and backend (?). 

This diverges quite a bit from the current way of managing permissions based on resource names. It seems the permissions would need extensive changes to allow that level of genericity. 

In the current format, the `permissions` are applied to one or multiple Kafka `clusters`. Its quite flexible since you can chose to have permissions for topicA in clusterA, but optionally not in clusterB. 

Whereas you may have multiple Kafka Connect clusters linked to a single Kafka cluster, and each Connect could need to have different permissions. 

If we were to simply add a kind of `resource: connector` to keep the permissions "flat", it would need an additional key to map it to a connect cluster.

```yaml
# Connector
        - resource: connector
          value: "my_connectors_.*" # regex to match connectors
          actions: [ view, edit, create, restart, reset_offsets ]
          connect: "my_connect_cluster" # could be a regex as well
```

In that case i'm not sure the `connect` resource is even needed since no operations are currently done on the cluster level, only on the connectors.

**Comment by Haarolean** [comment]
https://github.com/provectus/kafka-ui/issues/4253#issuecomment-1921311664

**Comment by Dugong42** [comment]
**Proposal:** to avoid changing the `permissions` object structure, the `connect` cluster could be included as a prefix inside the `value`, followed by the `connector`.

`value: "connect-name:connector-name"`

```yaml
# cluster 
rbac: 
  roles:
    - name: role-name
      ...
      permissions:
        - resource: connector
          # two-parts regex to match both the connect cluster and its connectors
          value: "my_connect_cluster:my_connectors_.*"
          actions: [ view, edit, create, restart, reset_offsets ]
```

**Comment by Haarolean** [comment]
> `permission[].value` would expect either one or two regexes separated by `:`
> 
> value: "my_connect_cluster:my_connectors_.*"
> * connect: `my_connect_cluster`
> * connector: `my_connectors_.*`
> 
> Omitting the second part would stay backwards compatible by matching with all connectors `.*`
> 
> value: "my_connect_cluster"
> * connect: `my_connect_cluster`
> * connector: `.*`

Unlikely, but what if the connector name contained a colon already?

**Comment by germanosin** [comment]
According to our RBAC model, a permission should only consist of the resource type, a list of actions, and a value (used as a regex for the name). Introducing additional fields would significantly alter the current logic. What we actually need is hierarchical permissions, and to support that, it would be more appropriate to add a parent field to the role entity.

**Comment by Dugong42** [comment]
> Unlikely, but what if the connector name contained a colon already?

Then we have a problem ^^'. Indeed that would not be compatible with such connector names.

>  What we actually need is hierarchical permissions

Got it. Both good remarks. I'll look into this.

**Comment by Dugong42** [comment]
@germanosin my previous proposal was to have a new `connector` resource with a reference to its parent as a `connect` field.

```yaml
        - resource: connector
          value: "my_connectors_.*" # regex to match connectors
          actions: [ view, edit, create, restart, reset_offsets ]
          connect: "my_connect_cluster" # could be a regex as well
```

The `connect` field:
- could be an optional field, regex matching the parent cluster(s) if specified, or matching all connect clusters otherwise.
- or it could be mandatory and only match a specific cluster in a more traditional hierarchical sense.

What's your take on this?

**Comment by germanosin** [comment]
> connect

@Dugong42
In my opinion, the "connect" field feels too specific and may introduce unnecessary complexity down the line, especially if we encounter other types of hierarchical entities. I would prefer using a more general parent field—either within the permission model or the role model.

**Comment by Dugong42** [comment]
Here we need to specify which `connect` cluster a `connector` permissions applies to, as the permission model currently only allows to target a resource by name, without knowledge of any metadata/context.

I believe that `role` composition/inheritance is a different topic, as it has more to do with the relationship between users and sets of permissions. It could allow for leaner rbac configs, but that a different issue.

Adding a `parent` field to the `permission` mode is better as it brings more context about the target resource. This may mean different things for a given resource, and still needs to be specifically implemented, but it would provide a way to both **identify the scope** and **inherit permissions** from the parent of a resource.

How's this?

```yaml
kafka:
  clusters:
    - name: kafka_1
      ...
      kafkaConnect:
        - name: connect_1
          ...
        - name: connect_2
          ...
rbac:
  roles:
    - name: role_name
      clusters: [ kafka_1 ]
      subjects: [ ... ]
      permissions:
        - resource: connect
          value: connect_1
          actions: [ view ]
        - resource: connect
          value: connect_2
          actions: [ view ]
        - resource: connector
          value: my_connectors_.* # regex
          actions: [ edit ]
          parent: connect_1
```
With this, permissions that apply to a `connect` cluster would also apply to its `connector` children.

> **Example:** A `user` granted `role_name` visiting the `kafka_1` tab would be allowed to `view` all connectors of `connect_1` and  `connect_2`, and `edit` connectors of `connect_1` matching `my_connectors_.*`.

**Comment by Dugong42** [comment]
Still considering alternatives. If we look at the cardinality of some services supported by kafbat, we have:

- kafka clusters [0..n]
  - topics [0..n]
  - consumer groups [0..n]
  - acls [0..n]
  - connect clusters [0..n]
    - connectors [0..n]
    - _loggers, plugins [0..n] (unsupported)_
  - schema registry [1]
    - schemas/subjects [0..n]
  - ksqldb [1]

A **connect cluster** is thus closer to a **kafka cluster** than to a **topic**, as it's a node that can hold different **resources**.
A `resource` should map uniquely to an actual object, that's the case for topics, consumers, acls because the `clusters` can be defined in the `role`. However for 

There is already a hierarchical relation between the `resources` and the `clusters` they apply to. The issue is that kafka connect is also a plural, so connectors

**Comment by Dugong42** [comment]
Currently the only available Kafbat operations for Connect are about `connectors`, however it's not excluded that more may be supported in the future.
Indeed it could be possible to display available plugins, change logger levels, or use any future Connect API.

When the following `connect` permission is set, it currently means that every connectors of cluster `connect_1` are readable, but it also serves to display the **Kafka Connect** tab:

```yaml
permissions:
        - resource: connect
          value: connect_1
          actions: [ view ]
```

This is equivalent to the following `connector` permission:

```yaml
permissions:
        - resource: connector
          value: .*
          actions: [ view ]
          parent: connect_1
```

For backwards compatibility, `connect` permissions should be merged with and thus override any `connector` permissions. As a result, if more Kafka Connect features ever get implemented, they may also require a specific `resource` to avoid overriding `connector` permissions..

**Comment by Dugong42** [comment]
I've found a few things to address about hierarchy.

> TLDR;
> - **Selection**: we should be able to unambiguously select a child resource, and we need the whole chain of parents for that
> - **Inheritance**: a child resource should inherit permissions from their parent(s) for backwards compatibility

## The issue with selecting a resource
As presented in a previous proposal, a strictly textual (i.e no regex) `parent` field on the child could be matched with the `value` of a parent, however it's not truly generic because it raises some questions if the resource has multiple levels of parents.

Example: `/myapp/mygroup/myresource`

`permissions` are not a 1 to 1 mapping to resources, and since the parent's value could be a pattern, a child may match multiple parents.

Resource hierarchy example:
- a1
  - b1
    - c1 
- a2
  - b1
    - c1 

If we try to set a permission for `a1.b1.c1` by specifying its direct parent `b1`, there's is some ambiguity as `b1` also exists under `a2`. That's a problem because we want resources to be uniquely defined.

So, to properly select a resource, we should be able to define all of its parents.

## Proposal

We could instead use an array of  `parents` or a recursive structure.:

```yaml
    permissions:
    - resource: typeC
      value: "c1"
      actions: [ VIEW ]
      parents:
      - resource: typeB
        value: b1 # pattern or string
      - resource: typeA
        value: a1 # pattern or string
```

Considering the order of parents in the array does not matter is a kinda bold assumption, even though currently ancestry is static. If we believe it could break at some point, then it's wiser to define parents recursively.

```yaml
    permissions:
    - resource: typeC
      value: "c1"
      actions: [ VIEW ]
      parent:
        resource: typeB
        value: b1
        parent: 
          resource: typeA
          value: a1
          parent: { ... } # optional
```

When accessing a1.b1.c1:
- the resource's parents must match any parent defined in the `permission` (will default to `.*` if undefined)
- permissions to the parents will be inherited.

That way we will not run into issues if there is ever more resource levels.

## Applied use case

Example with connectors:
```yaml
    permissions:
    - resource: connector
      value: "foo.*"
      actions: [ EDIT ]
      parent:
        resource: connect
        value: connect_1
    - resource: connect # optional, inherited by all its connectors
      value: connect_1
      actions: [ VIEW ]
```

@germanosin @Haarolean What do you think? Open to discussion/suggestions. :)

edit: promote the recursive structure

**Comment by shailendraBelieve** [comment]
Hello, any update on this issue ? It's quite complicated for us since we have one kafka-connect for several engineering teams and having broad permissions like that is dangerous

**Comment by Haarolean** [comment]
> Hello, any update on this issue ? It's quite complicated for us since we have one kafka-connect for several engineering teams and having broad permissions like that is dangerous

@shailendraBelieve you can sponsor this issue. Let me know if you'd like us to come up with an estimate for this.

**Comment by shailendraBelieve** [comment]
Let me check with my boss. An estimate would help me pitch the topic indeed.

**Comment by Dugong42** [comment]
Hi, I've been sporadically looking into how to implement this. I've made some progress on the backend, but haven't touched the front yet, and there's still quite some work to be done. 

I've opted to add a `connector` resource and an optional generic `parent` object to the `permission` model, as described above. It's recursive so as to allow any hierarchical resource structure.

To stay compatible to the existing permissions, this required to implement **permission inheritance**.

I found hard to do all this without introducing breaking changes or future headaches, as it requires a rather large rework of the authorization layer and deep changes to the permissions data model, which impacts many different parts (controller, audit logs, front).

See WIP: https://github.com/kafbat/kafka-ui/compare/main...Dugong42:kafka-ui:feat/connector-permissions

**Comment by Dugong42** [comment]
I also ran into some retro-compatibility issues in the front-end.

The CONNECT.VIEW permission is used to:
- Show the Kafka Connect tab
- Show and configure Kafka Connect in the Configuration wizard
- Show all connectors

With the new granular CONNECTOR permission, this becomes a problem since we need CONNECT.VIEW to show the Connect tab, which would also show all connectors, defeating the purpose. Doing some kind of "reverse" permission inheritance seems out of the question.

We could instead introduce new CONNECT actions (CLUSTER_VIEW and CLUSTER_EDIT) to configure access to the Connect cluster features (displaying the tab and the config wizard).
